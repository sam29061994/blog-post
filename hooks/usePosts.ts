import { useQuery } from 'react-query';

export type PostData = {
  title: string;
  description: string;
  createdAt: string;
  updatedAt: string;
  id: string;
  authors: [
    {
      name: string;
      id: string;
    }
  ];
  comments: [{ title: string; description: string; id: string }];
};

export const fetchPosts = async (): Promise<PostData[]> => {
  try {
    const response = await fetch(
      'https://6144e843411c860017d256f0.mockapi.io/api/v1/posts'
    );

    const result = await response.json();
    return result;
  } catch (e) {
    throw new Error((e as Error).message);
  }
};

export const usePosts = () => {
  const { data, error, isLoading, isError, isSuccess } = useQuery<PostData[]>(
    'posts',
    fetchPosts
  );
  return { data, error, isLoading, isError, isSuccess };
};
