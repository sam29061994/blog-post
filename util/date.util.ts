import moment from 'moment';

export const sortByRecentDate = (date1: Date, date2: Date) => {
  return getTime(date2) - getTime(date1);
};

export const getTime = (date?: Date) => {
  return date != null ? new Date(date).getTime() : 0;
};
export const convertDateToTimeAgo = (date: Date): string => {
  return moment(date).fromNow();
};
