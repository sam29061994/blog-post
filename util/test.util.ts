import { PostData } from '../hooks/usePosts';
export const post: PostData = {
  title: 'title1',
  description: 'description1',
  createdAt: '2021-05-20T01:13:07.861Z',
  updatedAt: '2021-05-20T01:13:07.861Z',
  id: '1',
  authors: [
    {
      name: 'sam',
      id: '1',
    },
  ],
  comments: [
    {
      title: 'comments1',
      description: 'description1',
      id: '1',
    },
  ],
} as PostData;
