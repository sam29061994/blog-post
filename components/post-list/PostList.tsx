import { useEffect } from 'react';
import { usePosts, PostData } from '../../hooks/usePosts';
import { Empty, List, message } from 'antd';
import PostListItem from '../post/PostListItem';
import { sortByRecentDate } from '../../util/date.util';
import * as CSS from 'csstype';

export type PostListProps = {
  pageSize: number;
};

const paginationStyle: CSS.Properties = {
  textAlign: 'left',
  padding: '10px 0',
};

const PostList = ({ pageSize }: PostListProps) => {
  const { data, isLoading, isError } = usePosts();

  useEffect(() => {
    if (isError) message.error('Error Fetching Posts');
  }, [isError]);

  const sortPosts = (first: PostData, second: PostData) =>
    sortByRecentDate(new Date(first.updatedAt), new Date(second.updatedAt));

  if (!data) return <Empty />;

  return (
    <List
      itemLayout="vertical"
      size="small"
      pagination={{
        pageSize: pageSize,
        defaultPageSize: 5,
        hideOnSinglePage: true,
        position: 'bottom',
        style: paginationStyle,
      }}
      dataSource={data.sort(sortPosts)}
      renderItem={(item) => (
        <PostListItem key={item.title} post={item} isLoading={isLoading} />
      )}
    />
  );
};

export default PostList;
