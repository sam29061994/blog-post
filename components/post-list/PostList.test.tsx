/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { render, cleanup, screen } from '@testing-library/react';
import PostList, { PostListProps } from './PostList';
import * as postModule from '../../hooks/usePosts';
import { post } from '../../util/test.util';

describe('PostListItemCard', () => {
  const props: PostListProps = {
    pageSize: 2,
  };

  beforeAll(() => {
    //Mocking methods which are not implemented in JSDOM
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });

  afterEach(() => {
    cleanup();
  });

  test('renders post list', () => {
    const spy = jest.spyOn(postModule, 'usePosts').mockImplementation(() => ({
      data: [post],
      error: null,
      isLoading: false,
      isError: false,
      isSuccess: false,
    }));

    render(<PostList {...props} />);
    expect(spy).toBeCalled();
    const title = screen.getByText(post.title);
    const description = screen.getByText(post.description);
    expect(title).toBeVisible();
    expect(description).toBeVisible();
  });

  test('renders No data Component', () => {
    const spy = jest.spyOn(postModule, 'usePosts').mockImplementation(() => ({
      data: [],
      error: null,
      isLoading: false,
      isError: false,
      isSuccess: false,
    }));

    render(<PostList {...props} />);
    expect(spy).toBeCalled();
    expect(screen.getByText('No Data')).toBeInTheDocument();
  });

  test('renders Error message', () => {
    const spy = jest.spyOn(postModule, 'usePosts').mockImplementation(() => ({
      data: [],
      error: null,
      isLoading: false,
      isError: true,
      isSuccess: false,
    }));

    render(<PostList {...props} />);
    expect(spy).toBeCalled();
    expect(screen.getByText('Error Fetching Posts')).toBeInTheDocument();
  });

  test('pagination works correctly', () => {
    const spy = jest.spyOn(postModule, 'usePosts').mockImplementation(() => ({
      data: [post, { ...post, title: 'title2' }, { ...post, title: 'title3' }],
      error: null,
      isLoading: false,
      isError: true,
      isSuccess: false,
    }));

    const { queryByRole, getByRole, container } = render(
      <PostList {...props} />
    );
    expect(spy).toBeCalled();
    expect(
      container.getElementsByClassName('ant-list-pagination').length
    ).toBeTruthy();
    expect(getByRole('listitem', { name: '2' })).toBeInTheDocument();
    expect(queryByRole('listitem', { name: '3' })).toBeNull();
  });

  test('matches snapshot', () => {
    const { asFragment } = render(<PostList {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
