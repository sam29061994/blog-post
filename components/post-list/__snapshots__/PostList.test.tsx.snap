// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostListItemCard matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ant-list ant-list-vertical ant-list-sm ant-list-split ant-list-something-after-last-item"
  >
    <div
      class="ant-spin-nested-loading"
    >
      <div
        class="ant-spin-container"
      >
        <ul
          class="ant-list-items"
        >
          <li
            class="ant-list-item ant-list-item-no-flex sc-bdvvtL bQcSwY"
          >
            <div
              class="ant-list-item-meta"
            >
              <div
                class="ant-list-item-meta-avatar"
              >
                <span
                  style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
                >
                  <span
                    style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
                  >
                    <img
                      alt=""
                      aria-hidden="true"
                      src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2740%27%20height=%2740%27/%3e"
                      style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                    />
                  </span>
                  <img
                    alt="Author Image"
                    data-nimg="intrinsic"
                    decoding="async"
                    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                  />
                  <noscript />
                </span>
              </div>
              <div
                class="ant-list-item-meta-content"
              >
                <h4
                  class="ant-list-item-meta-title"
                >
                  <a
                    href="title1"
                  >
                    title1
                  </a>
                </h4>
                <div
                  class="ant-list-item-meta-description"
                >
                  10 months ago
                </div>
              </div>
            </div>
            description1
            <ul
              class="ant-list-item-action"
            >
              <li>
                <div
                  class="ant-space ant-space-horizontal ant-space-align-center"
                >
                  <div
                    class="ant-space-item"
                  >
                    <span>
                      <span
                        aria-label="user"
                        class="anticon anticon-user"
                        role="img"
                        style="margin-right: 10px;"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="user"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M858.5 763.6a374 374 0 00-80.6-119.5 375.63 375.63 0 00-119.5-80.6c-.4-.2-.8-.3-1.2-.5C719.5 518 760 444.7 760 362c0-137-111-248-248-248S264 225 264 362c0 82.7 40.5 156 102.8 201.1-.4.2-.8.3-1.2.5-44.8 18.9-85 46-119.5 80.6a375.63 375.63 0 00-80.6 119.5A371.7 371.7 0 00136 901.8a8 8 0 008 8.2h60c4.4 0 7.9-3.5 8-7.8 2-77.2 33-149.5 87.8-204.3 56.7-56.7 132-87.9 212.2-87.9s155.5 31.2 212.2 87.9C779 752.7 810 825 812 902.2c.1 4.4 3.6 7.8 8 7.8h60a8 8 0 008-8.2c-1-47.8-10.9-94.3-29.5-138.2zM512 534c-45.9 0-89.1-17.9-121.6-50.4S340 407.9 340 362c0-45.9 17.9-89.1 50.4-121.6S466.1 190 512 190s89.1 17.9 121.6 50.4S684 316.1 684 362c0 45.9-17.9 89.1-50.4 121.6S557.9 534 512 534z"
                          />
                        </svg>
                      </span>
                      1
                    </span>
                  </div>
                </div>
                <em
                  class="ant-list-item-action-split"
                />
              </li>
              <li>
                <div
                  class="ant-space ant-space-horizontal ant-space-align-center"
                >
                  <div
                    class="ant-space-item"
                  >
                    <span>
                      <span
                        aria-label="message"
                        class="anticon anticon-message"
                        role="img"
                        style="margin-right: 10px;"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="message"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M464 512a48 48 0 1096 0 48 48 0 10-96 0zm200 0a48 48 0 1096 0 48 48 0 10-96 0zm-400 0a48 48 0 1096 0 48 48 0 10-96 0zm661.2-173.6c-22.6-53.7-55-101.9-96.3-143.3a444.35 444.35 0 00-143.3-96.3C630.6 75.7 572.2 64 512 64h-2c-60.6.3-119.3 12.3-174.5 35.9a445.35 445.35 0 00-142 96.5c-40.9 41.3-73 89.3-95.2 142.8-23 55.4-34.6 114.3-34.3 174.9A449.4 449.4 0 00112 714v152a46 46 0 0046 46h152.1A449.4 449.4 0 00510 960h2.1c59.9 0 118-11.6 172.7-34.3a444.48 444.48 0 00142.8-95.2c41.3-40.9 73.8-88.7 96.5-142 23.6-55.2 35.6-113.9 35.9-174.5.3-60.9-11.5-120-34.8-175.6zm-151.1 438C704 845.8 611 884 512 884h-1.7c-60.3-.3-120.2-15.3-173.1-43.5l-8.4-4.5H188V695.2l-4.5-8.4C155.3 633.9 140.3 574 140 513.7c-.4-99.7 37.7-193.3 107.6-263.8 69.8-70.5 163.1-109.5 262.8-109.9h1.7c50 0 98.5 9.7 144.2 28.9 44.6 18.7 84.6 45.6 119 80 34.3 34.3 61.3 74.4 80 119 19.4 46.2 29.1 95.2 28.9 145.8-.6 99.6-39.7 192.9-110.1 262.7z"
                          />
                        </svg>
                      </span>
                      1
                    </span>
                  </div>
                </div>
              </li>
            </ul>
          </li>
          <li
            class="ant-list-item ant-list-item-no-flex sc-bdvvtL bQcSwY"
          >
            <div
              class="ant-list-item-meta"
            >
              <div
                class="ant-list-item-meta-avatar"
              >
                <span
                  style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
                >
                  <span
                    style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
                  >
                    <img
                      alt=""
                      aria-hidden="true"
                      src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2740%27%20height=%2740%27/%3e"
                      style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                    />
                  </span>
                  <img
                    alt="Author Image"
                    data-nimg="intrinsic"
                    decoding="async"
                    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                  />
                  <noscript />
                </span>
              </div>
              <div
                class="ant-list-item-meta-content"
              >
                <h4
                  class="ant-list-item-meta-title"
                >
                  <a
                    href="title2"
                  >
                    title2
                  </a>
                </h4>
                <div
                  class="ant-list-item-meta-description"
                >
                  10 months ago
                </div>
              </div>
            </div>
            description1
            <ul
              class="ant-list-item-action"
            >
              <li>
                <div
                  class="ant-space ant-space-horizontal ant-space-align-center"
                >
                  <div
                    class="ant-space-item"
                  >
                    <span>
                      <span
                        aria-label="user"
                        class="anticon anticon-user"
                        role="img"
                        style="margin-right: 10px;"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="user"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M858.5 763.6a374 374 0 00-80.6-119.5 375.63 375.63 0 00-119.5-80.6c-.4-.2-.8-.3-1.2-.5C719.5 518 760 444.7 760 362c0-137-111-248-248-248S264 225 264 362c0 82.7 40.5 156 102.8 201.1-.4.2-.8.3-1.2.5-44.8 18.9-85 46-119.5 80.6a375.63 375.63 0 00-80.6 119.5A371.7 371.7 0 00136 901.8a8 8 0 008 8.2h60c4.4 0 7.9-3.5 8-7.8 2-77.2 33-149.5 87.8-204.3 56.7-56.7 132-87.9 212.2-87.9s155.5 31.2 212.2 87.9C779 752.7 810 825 812 902.2c.1 4.4 3.6 7.8 8 7.8h60a8 8 0 008-8.2c-1-47.8-10.9-94.3-29.5-138.2zM512 534c-45.9 0-89.1-17.9-121.6-50.4S340 407.9 340 362c0-45.9 17.9-89.1 50.4-121.6S466.1 190 512 190s89.1 17.9 121.6 50.4S684 316.1 684 362c0 45.9-17.9 89.1-50.4 121.6S557.9 534 512 534z"
                          />
                        </svg>
                      </span>
                      1
                    </span>
                  </div>
                </div>
                <em
                  class="ant-list-item-action-split"
                />
              </li>
              <li>
                <div
                  class="ant-space ant-space-horizontal ant-space-align-center"
                >
                  <div
                    class="ant-space-item"
                  >
                    <span>
                      <span
                        aria-label="message"
                        class="anticon anticon-message"
                        role="img"
                        style="margin-right: 10px;"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="message"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M464 512a48 48 0 1096 0 48 48 0 10-96 0zm200 0a48 48 0 1096 0 48 48 0 10-96 0zm-400 0a48 48 0 1096 0 48 48 0 10-96 0zm661.2-173.6c-22.6-53.7-55-101.9-96.3-143.3a444.35 444.35 0 00-143.3-96.3C630.6 75.7 572.2 64 512 64h-2c-60.6.3-119.3 12.3-174.5 35.9a445.35 445.35 0 00-142 96.5c-40.9 41.3-73 89.3-95.2 142.8-23 55.4-34.6 114.3-34.3 174.9A449.4 449.4 0 00112 714v152a46 46 0 0046 46h152.1A449.4 449.4 0 00510 960h2.1c59.9 0 118-11.6 172.7-34.3a444.48 444.48 0 00142.8-95.2c41.3-40.9 73.8-88.7 96.5-142 23.6-55.2 35.6-113.9 35.9-174.5.3-60.9-11.5-120-34.8-175.6zm-151.1 438C704 845.8 611 884 512 884h-1.7c-60.3-.3-120.2-15.3-173.1-43.5l-8.4-4.5H188V695.2l-4.5-8.4C155.3 633.9 140.3 574 140 513.7c-.4-99.7 37.7-193.3 107.6-263.8 69.8-70.5 163.1-109.5 262.8-109.9h1.7c50 0 98.5 9.7 144.2 28.9 44.6 18.7 84.6 45.6 119 80 34.3 34.3 61.3 74.4 80 119 19.4 46.2 29.1 95.2 28.9 145.8-.6 99.6-39.7 192.9-110.1 262.7z"
                          />
                        </svg>
                      </span>
                      1
                    </span>
                  </div>
                </div>
              </li>
            </ul>
          </li>
        </ul>
      </div>
    </div>
    <div
      class="ant-list-pagination"
    >
      <ul
        class="ant-pagination"
        style="text-align: left; padding: 10px 0px;"
        unselectable="unselectable"
      >
        <li
          aria-disabled="true"
          class="ant-pagination-prev ant-pagination-disabled"
          title="Previous Page"
        >
          <button
            class="ant-pagination-item-link"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <span
              aria-label="left"
              class="anticon anticon-left"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="left"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                />
              </svg>
            </span>
          </button>
        </li>
        <li
          class="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
          tabindex="0"
          title="1"
        >
          <a
            rel="nofollow"
          >
            1
          </a>
        </li>
        <li
          class="ant-pagination-item ant-pagination-item-2"
          tabindex="0"
          title="2"
        >
          <a
            rel="nofollow"
          >
            2
          </a>
        </li>
        <li
          aria-disabled="false"
          class="ant-pagination-next"
          tabindex="0"
          title="Next Page"
        >
          <button
            class="ant-pagination-item-link"
            tabindex="-1"
            type="button"
          >
            <span
              aria-label="right"
              class="anticon anticon-right"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="right"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                />
              </svg>
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
